src/honey_badger.rs

```diff
@@ -200,43 +222,213 @@ where
+    /// moves those transactions into a batch, outputs the batch and updates the epoch.
+    fn try_output_batch(&mut self) -> HoneyBadgerResult<bool> {
+        // Wait until selections have been successfully decoded for all proposer nodes with correct
+        // ciphertext outputs.
+        if !self.all_selections_decrypted() {
+            return Ok(false);
+        }
+
+        // Deserialize the output.
+        let transactions: BTreeMap<NodeUid, Vec<Tx>> = self
+            .decrypted_selections
+            .iter()
+            .flat_map(|(proposer_id, ser_batch)| {
+                // If serialization fails, the proposer of that batch is faulty. Ignore it.
+                bincode::deserialize::<Vec<Tx>>(&ser_batch)
+                    .ok()
+                    .map(|proposed| (proposer_id.clone(), proposed))
+            })
+            .collect();
+        let batch = Batch {
+            epoch: self.epoch,
+            transactions,
+        };
+        {
+            let tx_set: HashSet<&Tx> = batch.iter().collect();
+            // Remove the output transactions from our buffer.
+            self.buffer.retain(|tx| !tx_set.contains(&tx));
+        }
+        debug!(
+            "{:?} Epoch {} output {:?}",
+            self.netinfo.our_uid(),
+            self.epoch,
+            batch.transactions
+        );
+        // Queue the output and advance the epoch.
+        self.output.push_back(batch);
+        self.update_epoch()?;
+        Ok(true)
+    }
+
+    /// Increments the epoch number and clears any state that is local to the finished epoch.
+    fn update_epoch(&mut self) -> HoneyBadgerResult<()> {
+        // Clear the state of the old epoch.
+        self.ciphertexts.remove(&self.epoch);
+        self.decrypted_selections.clear();
+        self.received_shares.remove(&self.epoch);
+        self.epoch += 1;
+        // Handle any decryption shares received for the new epoch.
+        if !self.try_decrypt_and_output_batch()? {
+            // Continue with this epoch if a batch is not output by `try_decrypt_and_output_batch`.
+            self.propose()?;
+        }
+        Ok(())
+    }
+
+    /// Tries to decrypt transaction selections from all proposers and output those transactions in
+    /// a batch.
+    fn try_decrypt_and_output_batch(&mut self) -> HoneyBadgerResult<bool> {
+        let mut proposer_ids: Option<BTreeSet<NodeUid>> = None;
+        if let Some(shares) = self.received_shares.get(&self.epoch) {
+            proposer_ids = Some(shares.keys().cloned().collect());
+        }
+        if let Some(proposer_ids) = proposer_ids {
+            // Try to output a batch if there is a non-empty set of proposers for which we have already received
+            // decryption shares.
+            if !proposer_ids.is_empty()
+                && proposer_ids
+                    .iter()
+                    .all(|proposer_id| self.try_decrypt_proposer_selection(proposer_id))
```

Comment:
The criteria here should probably be:

* `CommonSubset` has already output in this epoch.
* For every proposer from `CommonSubset`'s output, we have received `f + 1` valid decryption shares.

If there are at most `f` faulty nodes, it should be guaranteed that that eventually happens _and_ that decryption succeeds.

Calling `try_decrypt_proposer_selection` every time looks like it might be expensive: wouldn't that repeatedly decrypt and overwrite the same thing? (Or should that method just return `true` immediately, if there's already an entry in `decrypted_selections`?)